File: LeetCode/Search Insert Position.py
class Solution:
    # @param A, a list of integers
    # @param target, an integer to be inserted
    # @return integer
    def searchInsert(self, A, target):
        left = 0
        right = len(A)
        while left <= right:
            mid = (right + left) >> 1
            # print (left, mid, right)

            if mid < len(A) and A[mid] > target:
                right = mid - 1
            elif mid < len(A) and A[mid] < target:
                left = mid + 1
            else:
                return mid

        if A[mid] >= target:
            return mid
        else:
            return mid+1

if __name__ == '__main__':
    s = Solution()
    A = [1,3,5,6]
    k = 12
    print s.searchInsert(A,k)

File: LeetCode/Search Insert Position.cpp
#include <iostream>
using namespace std;

class Solution {
public:
    int searchInsert(int A[], int n, int target) {
        int m, l = 0, r = n-1;
        while (l <= r){
            m = (r - l)/2 + l;
            if (A[m] < target){
                l = m + 1;
            }else if (A[m] > target){
                r = m - 1;
            }else{
                return m;
            }
        }
        if (target > A[m])
            return m + 1;
        else
            return m;
    }
};

int main(int argc, char const *argv[])
{
    Solution sol;
    int A[] = {2, 4, 6, 7, 10};
    cout << sol.searchInsert(A, 5, 11) << endl;
    return 0;
}